text/0000-closure-lifetime-binder.md

```diff
@@ -0,0 +1,279 @@
+
+```rust
+fn func(_: impl Fn(&i32) -> &i32) {}
+
+fn main() {
+    func(|arg: &i32| -> &i32 { arg });
+}
+```
+
+Notice that we've *elided* the lifetime in `&i32`. When a lifetime is written this way, Rust will infer its value based on how it's used.
+
+In this case, our closure needs to be able to accept an `&i32` with *any* lifetime. This is because our closure needs to implement `Fn(&i32) -> &i32` - this is syntactic sugar for `for<'a> Fn(&'a i32) -> &'a i32`.
+
+We can make this explicit by writing our closure in the following way:
+
+```rust
+fn func(_: impl Fn(&i32) -> &i32) {}
+
+fn main() {
+    func(for<'a> |arg: &'a i32| -> &'a i32 { arg });
+}
+```
+
+This indicates to both the compiler and the user that this closure can accept an `&i32` with *any* lifetime, and returns an `&i32` with the same lifetime.
+
+However, there are cases where a closure *cannot* accept any lifetime - it can only accept some particular lifetime. Consider the following code:
+
+```rust
+fn main() {
+    let mut values: Vec<&bool> = Vec::new();
+    let first = true;
+    values.push(&first);
+
+    let mut closure = |value| values.push(value);
+    let second = false;
+    closure(&second);
+}
+```
+
+In this code, `closure` takes in an `&bool`, and pushes it to `values`. However, `closure` *cannot* accept an `&bool` with *any* lifetime - it can only work with some specific lifetime. To see this, consider this slight modification of the program:
+
+```rust
+fn main() {
+    let mut values: Vec<&bool> = Vec::new();
+    let first = true;
+    values.push(&first);
+
+    let mut closure = |value| values.push(value);
+    { // This new scope was added
+        let second = false;
+        closure(&second);
+    } // The scope ends here, causing `second` to be dropped
+    println!("Values: {:?}", values);
+}
+```
+
+This program fails to compile:
+
+```
+error[E0597]: `second` does not live long enough
+  --> src/main.rs:9:17
+   |
+9  |         closure(&second);
+   |                 ^^^^^^^ borrowed value does not live long enough
+10 |     }
+   |     - `second` dropped here while still borrowed
+11 |     println!("Values: {:?}", values);
+   |                              ------ borrow later used here
+```
+
+This is because `closure` can only accept an `&bool` with a lifetime that lives at least as long as `values`. If this code were to compile (that is, if `closure` could accept a `&bool` with the shorter lifetime associated with `&second`), then `values` would end up containing a reference to the freed stack variable `second`.
+
+Since `closure` cannot accept *any* lifetime, it cannot be written as `for<'a> |value: &'a bool| values.push(value)`. It's natural to ask - how *can* we write down an explicit lifetime for `value: &bool`?
+
+Unfortunately, Rust does not currently allow the signature of such a closure to be written explicitly. Instead, you must rely on type inference to choose the correct lifetime for you.
+
+# Reference-level explanation
```

Comment:
I am wondering if it makes sense to use this RFC as an opportunity to better document how the inference works today. Perhaps it would be better to do that as PRs on the Rust reference.